File: leetcode/155. Min Stack/MinStack.go
type MinStack struct {
	stack    []int
	minStack []int
}

func Constructor() MinStack {
	min := make([]int, 0)
	stack := make([]int, 0)
	minStack := MinStack{stack, min}
	return minStack
}

func (this *MinStack) Push(val int) {
	this.stack = append(this.stack, val)
	if len(this.minStack) == 0 {
		this.minStack = append(this.minStack, val)
	} else if val < this.minStack[len(this.minStack)-1] {
		this.minStack = append(this.minStack, val)
	} else {
		this.minStack = append(this.minStack, this.minStack[len(this.minStack)-1])
	}
}

func (this *MinStack) Pop() {
	this.stack = this.stack[:len(this.stack)-1]
	this.minStack = this.minStack[:len(this.minStack)-1]
}

func (this *MinStack) Top() int {
	top := this.stack[len(this.stack)-1]
	return top
}

func (this *MinStack) GetMin() int {
	min := this.minStack[len(this.minStack)-1]
	return min
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: leetcode/155. Min Stack/MinStack.cpp
class MinStack
{
public:
    vector<int> stack;
    vector<int> minStack;

    MinStack()
    {
    }

    void push(int val)
    {
        stack.push_back(val);
        if (minStack.size() == 0)
        {
            minStack.push_back(val);
        }
        else if (val < minStack[minStack.size() - 1])
        {
            minStack.push_back(val);
        }
        else
        {
            minStack.push_back(minStack[minStack.size() - 1]);
        }
    }

    void pop()
    {
        if (stack.size() > 0)
        {
            stack.pop_back();
            minStack.pop_back();
        }
    }

    int top()
    {
        if (stack.size() > 0)
        {
            return stack[stack.size() - 1];
        }
        return -1;
    }

    int getMin()
    {
        if (minStack.size() > 0)
        {
            return minStack[minStack.size() - 1];
        }
        return -1;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

//*******************************************

class MinStack
{
public:
    stack<int> s, s_min;
    MinStack()
    {
    }

    void push(int val)
    {
        s.emplace(val);
        if (s_min.empty())
        {
            s_min.emplace(val);
        }
        else if (s_min.top() > val)
        {
            s_min.emplace(val);
        }
        else
        {
            s_min.emplace(s_min.top());
        }
    }

    void pop()
    {
        if (!s_min.empty() && !s.empty())
        {
            s.pop();
            s_min.pop();
        }
    }

    int top()
    {
        int top = 0;
        if (!s.empty())
        {
            top = s.top();
        }
        return top;
    }

    int getMin()
    {
        int min = 0;
        if (!s_min.empty())
        {
            min = s_min.top();
        }
        return min;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */